lib/model/QubitDigitalObject.php

```diff
@@ -2764,8 +2800,70 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
             return false;
         }
 
-        $command = 'ffmpeg -y -i '.$originalPath.' -ar 44100 -c:v libx264 -pix_fmt yuv420p -c:a aac -movflags +faststart '.$newPath.' 2>&1';
-        exec($command, $output, $status);
+        // Read stream information
+        $format = self::readStreamInformation($originalPath);
+
+        error_log('Stream information: ' . json_encode($format, JSON_PRETTY_PRINT));
+
+        $videoCodec = isset($format['video']['codec_name']) ? $format['video']['codec_name'] : null;
+        $pixelFormat = isset($format['video']['pix_fmt']) ? $format['video']['pix_fmt'] : null;
+        $audioCodec = isset($format['audio']['codec_name']) ? $format['audio']['codec_name'] : null;
+        $sampleRate = isset($format['audio']['sample_rate']) ? $format['audio']['sample_rate'] : null;
+        $format_name = isset($format['format_name']) ? $format['format_name'] : null;
+
+        $reencodeVideo = ($videoCodec !== 'h264' || $pixelFormat !== 'yuv420p');
+        $reencodeAudio = ($audioCodec !== 'aac' || $sampleRate !== 44100);
+        $reformatFile = strpos($format_name, 'mp4') === false || strtolower(pathinfo($originalPath, PATHINFO_EXTENSION)) !== 'mp4';
+
+        $needFastStart = !self::isFastStarted($originalPath);
```

Comment:
I think you can short-circuit this check here to skip running the `ffprobe` command again in certain situations:

```suggestion
        $needFastStart = $reencodeVideo || !self::isFastStarted($originalPath);
```

Since if you're re-encoding the video, you'll want to faststart it by default.

Reply:
I believe it is possible for some MP4 files to be faststarted already, but do not meet the video encoding criteria. For e.g., here is a MP4 file that is already faststarted and has `hevc` video encoding:


```bash
╰─ ffprobe -v quiet -show_entries format=format_name -show_streams -print_format json invalid_vid_codec_faststarted.mp4 2>&1
{
    "streams": [
        {
            "index": 0,
            "codec_name": "hevc",
            "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
            "profile": "Main",
            "codec_type": "video",
            "codec_tag_string": "hev1",
            "codec_tag": "0x31766568",
            "width": 480,
            "height": 270,
            "coded_width": 480,
            "coded_height": 272,
            "closed_captions": 0,
            "film_grain": 0,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p",
            "level": 63,
            "color_range": "tv",
            "chroma_location": "left",
            "field_order": "progressive",
            "refs": 1,
            "view_ids_available": "",
            "view_pos_available": "",
            "id": "0x1",
            "r_frame_rate": "30/1",
            "avg_frame_rate": "3375/113",
            "time_base": "1/15360",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 461824,
            "duration": "30.066667",
            "bit_rate": "27420",
            "nb_frames": "900",
            "extradata_size": 2466,
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "non_diegetic": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0,
                "multilayer": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "VideoHandler",
                "vendor_id": "[0][0][0][0]",
                "encoder": "Lavc61.19.100 libx265"
            }
        },
        {
            "index": 1,
            "codec_name": "aac",
            "codec_long_name": "AAC (Advanced Audio Coding)",
            "profile": "LC",
            "codec_type": "audio",
            "codec_tag_string": "mp4a",
            "codec_tag": "0x6134706d",
            "sample_fmt": "fltp",
            "sample_rate": "44100",
            "channels": 2,
            "channel_layout": "stereo",
            "bits_per_sample": 0,
            "initial_padding": 0,
            "id": "0x2",
            "r_frame_rate": "0/0",
            "avg_frame_rate": "0/0",
            "time_base": "1/44100",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 1348578,
            "duration": "30.580000",
            "bit_rate": "2093",
            "nb_frames": "1318",
            "extradata_size": 5,
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "non_diegetic": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0,
                "multilayer": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "SoundHandler",
                "vendor_id": "[0][0][0][0]"
            }
        }
    ],
    "format": {
        "format_name": "mov,mp4,m4a,3gp,3g2,mj2"
    }
}

╭─    ~/Desktop/vid_test/valid ··························· ✔  09:03:12 
╰─ ffprobe -v trace invalid_vid_codec_faststarted.mp4 2>&1 | grep -E "type:'(mdat|moov)'"
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x11fe04e50] type:'moov' parent:'root' sz: 37421 36 148532
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x11fe04e50] type:'mdat' parent:'root' sz: 111075 37465 148532

```

Reply:
That makes sense, disregard my comment then! Still just looking at small ways to make efficiencies, looks like we're approaching a state where it's as efficient as possible. I would love if there were a way to output the moov/mdat atom information when I run FFprobe the first time but I don't think that's possible.